GeoIP backend - Hit on service's default network matching poisons the cache

- Program: Authoritative
 - Issue type: Bug report

### Short description

When using the GeoIP backend, as described on https://doc.powerdns.com/md/authoritative/backend-geoip/, we can configure services with subnet matching rules and a default matching rule.

I can send queries from all the configured subnets and I always get the correct answer.
If we send a query that matches no subnet, it returns the default value configured for the service, which is normal.

But then, after receiving that "default matched" query, it always answer with that value for everyone, for 20 seconds (query-cache-ttl delay I guess), even for clients that should be matched against one of the configured subnet.

### Environment
 - Operating system: Docker
 - Software version: PowerDNS Authoritative Server 4.1.1
 - Software source: PowerDNS repository

### Configuration

Command line : --config-dir=/powerdns/ --resolver=1.2.3.4:53

/powerdns/pdns.conf :
```
expand-alias=yes
launch=geoip
geoip-zones-file=/powerdns/zones.yml

```
/powerdns/zones.yml
```
domains:
- domain: example.com
  ttl: 30
  records:
    example.com:
      - soa: ns-aaa.example.com dns.example.com 2018050100 7200 3600 1209600 300
      - ns: ns-aaa.example.com
      - ns: ns-bbb.example.com
    ns-aaa.example.com:
      - a: 10.1.0.1
    ns-bbb.example.com:
      - a: 10.2.0.1
    foo-aaa.example.com:
      - alias: aaa.another.domain.com
    foo-bbb.example.com:
      - alias: bbb.another.domain.com
  services:
    foo.example.com:
      default: 'foo-aaa.example.com'
      10.1.0.0/16: 'foo-aaa.example.com'
      10.2.0.0/16: 'foo-bbb.example.com'
```

### Expected behaviour

The service's subnet mapping should always be respected.
The query cache should be involved only after resolving the service based on the subnet.

### Actual behaviour

Query Cache is filled only when hitting the default service mapping, and the cached value is then wrongly used for the duration of query-cache-ttl.

### Usecase

It's the whole point of the GeoIP backend to vary the answered value based on the client's IP address.
One query from an unknown IP should not poison the cache.

###

I will try to look into the implementation code if I have some time.
Thanks already.

I adapted the Dockerfile to build the geoip branch.

However I get stuck further down the compilation with these issues : 
```
pollmplexer.cc:11:52: error: no 'FDMultiplexer* FDMultiplexer::getMultiplexerSilent()' member function declared in class 'FDMultiplexer'
 FDMultiplexer* FDMultiplexer::getMultiplexerSilent()
                                                    ^
pollmplexer.cc:38:16: error: 'virtual void PollFDMultiplexer::getAvailableFDs(std::vector<int>&, int)' marked 'override', but does not override
   virtual void getAvailableFDs(std::vector<int>& fds, int timeout) override;
                ^~~~~~~~~~~~~~~
pollmplexer.cc:43:10: error: 'std::__cxx11::string PollFDMultiplexer::getName() const' marked 'override', but does not override
   string getName() const override
          ^~~~~~~
pollmplexer.cc: In function 'FDMultiplexer* make()':
pollmplexer.cc:53:32: error: invalid new-expression of abstract class type 'PollFDMultiplexer'
   return new PollFDMultiplexer();
                                ^
pollmplexer.cc:28:7: note:   because the following virtual functions are pure within 'PollFDMultiplexer':
 class PollFDMultiplexer : public FDMultiplexer
       ^~~~~~~~~~~~~~~~~
In file included from pollmplexer.cc:4:0:
mplexer.hh:133:23: note:        virtual std::__cxx11::string FDMultiplexer::getName()
   virtual std::string getName() = 0;
                       ^~~~~~~
```

If you have a working docker image or amd64 binaries/libs (with geoip) that I can use it'd be awesome.
Otherwise I'll continue to work on this build.

;-)

--- a/modules/geoipbackend/geoipbackend.cc
+++ b/modules/geoipbackend/geoipbackend.cc
@@ -406,7 +406,7 @@ bool GeoIPBackend::get(DNSResourceRecord &r) {
   return true;
 }
 
-string queryGeoIP(const string &ip, bool v6, GeoIPInterface::GeoIPQueryAttribute attribute, GeoIPNetmask& gl) {
+static string queryGeoIP(const string &ip, bool v6, GeoIPInterface::GeoIPQueryAttribute attribute, GeoIPNetmask& gl) {
   string ret = "unknown";
 
   for(auto const& gi: s_geoip_files) {

--- a/modules/geoipbackend/geoipbackend.cc
+++ b/modules/geoipbackend/geoipbackend.cc
@@ -35,13 +35,11 @@
 pthread_rwlock_t GeoIPBackend::s_state_lock=PTHREAD_RWLOCK_INITIALIZER;
 
 struct GeoIPDNSResourceRecord: DNSResourceRecord {
-public:
   int weight;
   bool has_weight;
 };
 
-class GeoIPDomain {
-public:
+struct GeoIPDomain {
   int id;
   DNSName domain;
   int ttl;

--- a/modules/geoipbackend/geoipbackend.cc
+++ b/modules/geoipbackend/geoipbackend.cc
@@ -301,7 +301,7 @@ GeoIPBackend::~GeoIPBackend() {
 }
 
 bool GeoIPBackend::lookup_static(const GeoIPDomain &dom, const DNSName &search, const QType &qtype, const DNSName& qdomain, const std::string &ip, GeoIPNetmask &gl, bool v6) {
-  const auto i = dom.records.find(search);
+  const auto& i = dom.records.find(search);
   int cumul_probability = 0;
   int probability_rnd = 1+(dns_random(1000)); // setting probability=0 means it never is used
 
@@ -334,23 +334,21 @@ bool GeoIPBackend::lookup_static(const GeoIPDomain &dom, const DNSName &search, 
 
 void GeoIPBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket *pkt_p, int zoneId) {
   ReadLock rl(&s_state_lock);
-  GeoIPDomain dom;
+  const GeoIPDomain* dom;
   GeoIPNetmask gl;
   bool found = false;
 
   if (d_result.size()>0)
     throw PDNSException("Cannot perform lookup while another is running");
 
-  DNSName search = qdomain;
-
   d_result.clear();
 
   if (zoneId > -1 && zoneId < static_cast<int>(s_domains.size()))
-    dom = s_domains[zoneId];
+    dom = &(s_domains[zoneId]);
   else {
     for(const GeoIPDomain& i : s_domains) {   // this is arguably wrong, we should probably find the most specific match
-      if (search.isPartOf(i.domain)) {
-        dom = i;
+      if (qdomain.isPartOf(i.domain)) {
+        dom = &i;
         found = true;
         break;
       }
@@ -367,10 +365,10 @@ void GeoIPBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket 
 
   gl.netmask = 0;
 
-  (void)this->lookup_static(dom, search, qtype, qdomain, ip, gl, v6);
+  (void)this->lookup_static(*dom, qdomain, qtype, qdomain, ip, gl, v6);
 
-  auto target = dom.services.find(search);
-  if (target == dom.services.end()) return; // no hit
+  const auto& target = (*dom).services.find(qdomain);
+  if (target == (*dom).services.end()) return; // no hit
 
   const NetmaskTree<vector<string> >::node_type* node = target->second.masks.lookup(ComboAddress(ip));
   if (node == NULL) return; // no hit, again.
@@ -400,7 +398,7 @@ void GeoIPBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket 
     sformat = DNSName(format2str(*it, ip, v6, gl));
 
     // see if the record can be found
-    if (this->lookup_static(dom, sformat, qtype, qdomain, ip, gl, v6))
+    if (this->lookup_static((*dom), sformat, qtype, qdomain, ip, gl, v6))
       return;
   }
 
@@ -417,12 +415,12 @@ void GeoIPBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket 
   if (!(qtype == QType::ANY || qtype == QType::CNAME)) return;
 
   DNSResourceRecord rr;
-  rr.domain_id = dom.id;
+  rr.domain_id = dom->id;
   rr.qtype = QType::CNAME;
   rr.qname = qdomain;
   rr.content = sformat.toString();
   rr.auth = 1;
-  rr.ttl = dom.ttl;
+  rr.ttl = dom->ttl;
   rr.scopeMask = gl.netmask;
   d_result.push_back(rr);
 }

--- a/modules/geoipbackend/geoipbackend.cc
+++ b/modules/geoipbackend/geoipbackend.cc
@@ -277,7 +277,7 @@ void GeoIPBackend::initialize() {
       }
     }
 
-    tmp_domains.push_back(dom);
+    tmp_domains.push_back(std::move(dom));
   }
 
   s_domains.clear();

--- a/modules/geoipbackend/geoipbackend.cc
+++ b/modules/geoipbackend/geoipbackend.cc
@@ -41,6 +41,8 @@ struct GeoIPDNSResourceRecord: DNSResourceRecord {
 
 struct GeoIPService {
   NetmaskTree<vector<string> > masks;
+  unsigned int netmask4;
+  unsigned int netmask6;
 };
 
 struct GeoIPDomain {
@@ -87,6 +89,7 @@ GeoIPBackend::GeoIPBackend(const string& suffix) {
 static vector<std::unique_ptr<GeoIPInterface> > s_geoip_files;
 
 string getGeoForLua(const std::string& ip, int qaint);
+static string queryGeoIP(const string &ip, bool v6, GeoIPInterface::GeoIPQueryAttribute attribute, GeoIPNetmask& gl);
 
 void GeoIPBackend::initialize() {
   YAML::Node config;
@@ -166,6 +169,8 @@ void GeoIPBackend::initialize() {
     }
 
     for(YAML::const_iterator service = domain["services"].begin(); service != domain["services"].end(); service++) {
+      unsigned int netmask4 = 0, netmask6 = 0;
+      DNSName srvName{service->first.as<string>()};
       NetmaskTree<vector<string> > nmt;
 
       // if it's an another map, we need to iterate it again, otherwise we just add two root entries.
@@ -181,7 +186,12 @@ void GeoIPBackend::initialize() {
             nmt.insert(Netmask("0.0.0.0/0")).second.assign(value.begin(),value.end());
             nmt.insert(Netmask("::/0")).second.swap(value);
           } else {
-            nmt.insert(Netmask(net->first.as<string>())).second.swap(value);
+            Netmask nm{net->first.as<string>()};
+            nmt.insert(nm).second.swap(value);
+            if (nm.isIpv6() == true && netmask6 < nm.getBits())
+              netmask6 = nm.getBits();
+            if (nm.isIpv6() == false && netmask4 < nm.getBits())
+              netmask4 = nm.getBits();
           }
         }
       } else {
@@ -194,7 +204,10 @@ void GeoIPBackend::initialize() {
         nmt.insert(Netmask("0.0.0.0/0")).second.assign(value.begin(),value.end());
         nmt.insert(Netmask("::/0")).second.swap(value);
       }
-      dom.services[DNSName(service->first.as<string>())].masks.swap(nmt);
+
+      dom.services[srvName].netmask4 = netmask4;
+      dom.services[srvName].netmask6 = netmask6;
+      dom.services[srvName].masks.swap(nmt);
     }
 
     // rectify the zone, first static records
@@ -364,6 +377,23 @@ void GeoIPBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket 
 
   DNSName sformat;
   gl.netmask = node->first.getBits();
+  // figure out smallest sensible netmask
+  if (gl.netmask == 0) {
+    GeoIPNetmask tmp_gl;
+    tmp_gl.netmask = 0;
+    // get netmask from geoip backend
+    if (queryGeoIP(ip, v6, GeoIPInterface::Name, tmp_gl) == "unknown") {
+      if (v6)
+        gl.netmask = target->second.netmask6;
+      else
+        gl.netmask = target->second.netmask4;
+    }
+  } else {
+    if (v6)
+      gl.netmask = target->second.netmask6;
+    else
+      gl.netmask = target->second.netmask4;
+  }
 
   // note that this means the array format won't work with indirect
   for(auto it = node->second.begin(); it != node->second.end(); it++) {

--- a/modules/geoipbackend/geoipbackend.cc
+++ b/modules/geoipbackend/geoipbackend.cc
@@ -39,11 +39,15 @@ struct GeoIPDNSResourceRecord: DNSResourceRecord {
   bool has_weight;
 };
 
+struct GeoIPService {
+  NetmaskTree<vector<string> > masks;
+};
+
 struct GeoIPDomain {
   int id;
   DNSName domain;
   int ttl;
-  map<DNSName, NetmaskTree<vector<string> > > services;
+  map<DNSName, GeoIPService> services;
   map<DNSName, vector<GeoIPDNSResourceRecord> > records;
 };
 
@@ -190,7 +194,7 @@ void GeoIPBackend::initialize() {
         nmt.insert(Netmask("0.0.0.0/0")).second.assign(value.begin(),value.end());
         nmt.insert(Netmask("::/0")).second.swap(value);
       }
-      dom.services[DNSName(service->first.as<string>())].swap(nmt);
+      dom.services[DNSName(service->first.as<string>())].masks.swap(nmt);
     }
 
     // rectify the zone, first static records
@@ -355,7 +359,7 @@ void GeoIPBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket 
   auto target = dom.services.find(search);
   if (target == dom.services.end()) return; // no hit
 
-  const NetmaskTree<vector<string> >::node_type* node = target->second.lookup(ComboAddress(ip));
+  const NetmaskTree<vector<string> >::node_type* node = target->second.masks.lookup(ComboAddress(ip));
   if (node == NULL) return; // no hit, again.
 
   DNSName sformat;

